subsys/task_wdt/task_wdt.c

```diff
@@ -0,0 +1,210 @@
+
+/*
+ * This dummy channel is used to continue feeding the hardware watchdog if the
+ * task watchdog timeouts are too long for regular updates
+ */
+#define TASK_WDT_BACKGROUND_CHANNEL (-1)
+
+/*
+ * Task watchdog channel data
+ */
+struct task_wdt_channel {
+	/* period in milliseconds used to reset the timeout, set to 0 to
+	 * indicate that the channel is available
+	 */
+	uint32_t reload_period;
+	/* abs. ticks when this channel expires (updated by task_wdt_feed) */
+	int64_t timeout_abs_ticks;
+	/* user data passed to the callback function */
+	void *user_data;
+	/* function to be called when watchdog timer expired */
+	task_wdt_callback_t callback;
+};
+
+/* array of all task watchdog channels */
+static struct task_wdt_channel channels[CONFIG_TASK_WDT_CHANNELS];
+
+/* timer used for watchdog handling */
+static struct k_timer timer;
+
+#ifdef CONFIG_TASK_WDT_HW_FALLBACK
+/* pointer to the hardware watchdog used as a fallback */
+static const struct device *hw_wdt_dev;
+static int hw_wdt_channel;
+static bool hw_wdt_started;
+#endif
+
+/**
+ * @brief Task watchdog timer callback.
+ *
+ * If the device operates as intended, this function will never be called,
+ * as the timer is continuously restarted with the next due timeout in the
+ * task_wdt_feed() function.
+ *
+ * If all task watchdogs have longer timeouts than the hardware watchdog,
+ * this function is called regularly (via the background channel). This
+ * should be avoided by setting CONFIG_TASK_WDT_MIN_TIMEOUT to the minimum
+ * task watchdog timeout used in the application.
+ *
+ * @param timer_id Pointer to the timer which called the function
+ */
+static void task_wdt_trigger(struct k_timer *timer_id)
+{
+	int channel_id = (int)k_timer_user_data_get(timer_id);
+
+#ifdef CONFIG_TASK_WDT_HW_FALLBACK
+	if (channel_id == TASK_WDT_BACKGROUND_CHANNEL) {
+		if (hw_wdt_dev) {
+			wdt_feed(hw_wdt_dev, 0);
+		}
+		return;
+	}
+#endif
+
+	if (channels[channel_id].reload_period == 0) {
+		/* channel was deleted */
+		return;
+	} else if (channels[channel_id].callback) {
+		channels[channel_id].callback(channel_id,
+			channels[channel_id].user_data);
+	} else {
+		sys_reboot(SYS_REBOOT_COLD);
+	}
+}
+
+int task_wdt_init(const struct device *hw_wdt)
+{
+	if (hw_wdt) {
+#ifdef CONFIG_TASK_WDT_HW_FALLBACK
+		struct wdt_timeout_cfg wdt_config;
+
+		/* set the hardware watchdog timeout 1 ms longer than the
+		 * minimum task wdt resolution to use it as fallback reset
+		 */
+		wdt_config.flags = WDT_FLAG_RESET_SOC;
+		wdt_config.window.min = 0U;
+		wdt_config.window.max = CONFIG_TASK_WDT_MIN_TIMEOUT + 1;
+		wdt_config.callback = NULL;
+
+		hw_wdt_dev = hw_wdt;
+		hw_wdt_channel = wdt_install_timeout(hw_wdt_dev, &wdt_config);
+#else
+		return -ENOTSUP;
+#endif
+	}
+
+	k_timer_init(&timer, task_wdt_trigger, NULL);
+
+	return 0;
+}
+
+int task_wdt_add(uint32_t reload_period, task_wdt_callback_t callback,
+		 void *user_data)
+{
+	/* look for unused channel (reload_period set to 0) */
```

Comment:
I think it's better to refuse "reload_period == 0" calls. It indicates invalid task_wdt, so there is no meaning for searching.

Reply:
Agree with @katsuster. Other than that, the addition of this subsys LGTM.

Reply:
I agree aswell. Fixed.